nixos/modules: fix systemd start rate-limits

lf- opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
These have been broken since 2016:
https://github.com/systemd/systemd/commit/f0367da7d1a61ad698a55d17b5c28ddce0dc265a
since StartLimitIntervalSec got moved into [Unit] from [Service].
StartLimitBurst has also been moved accordingly, so let's fix that one
too by adding an option ourselves.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) -- **unfortunately my computer can't run kvm so I was unable to run tests**
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

lf- committed:
nixos/modules: fix systemd start rate-limits

These were broken since 2016:
https://github.com/systemd/systemd/commit/f0367da7d1a61ad698a55d17b5c28ddce0dc265a
since StartLimitIntervalSec got moved into [Unit] from [Service].
StartLimitBurst has also been moved accordingly, so let's fix that one
too.

NixOS systems have been producing logs such as:
/nix/store/wf98r55aszi1bkmln1lvdbp7znsfr70i-unit-caddy.service/caddy.service:31:
Unknown key name 'StartLimitIntervalSec' in section 'Service', ignoring.

I have also removed some unnecessary duplication in units disabling
rate limiting since setting either interval or burst to zero disables it
(https://github.com/systemd/systemd/blob/ad16158c10dfc3258831a9ff2f1a988214f51653/src/basic/ratelimit.c#L16)

lf- commented:
oopsie, fixed an ofborg eval error. might work now

nh2 commented:
This contributes to #45785, right?

Independently, could you confirm whether

> These have been broken since 2016

is really true? In https://github.com/NixOS/nixpkgs/issues/45785#issuecomment-612268015 I found that systemd did not remove `serviceConfig.StartLimitIntervalSec` -- its legacy usage is still supported, but undocumented. But I don't know if that's really the case for the burst option. As the other one is not documented, you may not get the answer from the documentation, but only from the code (or trying it).

lf- commented:
I believe that's incorrect: StartLimitInterval is likely still accepted in Service (not 100% sure) but StartLimitIntervalSec is definitely not, at least according to my log lines warning on it (see the commit message and the caddy PR). StartLimitBurst I think *is* accepted there but I was touching the files anyway and it was also moved by upstream, so I fixed it as well. 

Regardless, we were relying on undocumented deprecated behaviour and I wanted to fix all of it now that I noticed it.

lf- commented:
> This contributes to #45785, right?

I guess so! This fixes #45785, arguably.

nh2 commented:
> but StartLimitIntervalSec is definitely not

@lf- You're right, I had typod it in the linked comment and written `StartLimitIntervalSec` twice (thus making it inconsistent with the nixpkgs code I quoted there). I've fixed it now.

nh2 reviewed (approved):
> I guess so! This fixes #45785, arguably.

Great! +1 for me on the change (but I cannot test the change to each of the services, so just +1'ing the approach).

Optionally: Would you also be up for implementing this note from #45785 that makes any further incorrect use of the deprecated field an evaluation warning?

> * [ ] We should probably also try and make usage of `StartLimitInterval` in the old section emit a warning when it's used, as having it in the new section has [benefits](https://github.com/NixOS/nixpkgs/issues/45786#issuecomment-440091879)

lf- commented:
@nh2 unfortunately I'm quite busy this next couple of weeks and probably don't have time to figure out how to do it. I doubt it's complex but I'd have to do a bit of research.

lf- committed:
nixos/modules: deprecation warning for StartLimitInterval in [Service]

This implements
https://github.com/NixOS/nixpkgs/issues/45786#issuecomment-440091879

lf- commented:
@nh2 I've implemented the deprecation warning as requested in 039200b.

nh2 commented on nixos/modules/system/boot/systemd.nix:
```diff
@@ -888,9 +890,13 @@ in
       "Service '${name}.service' with 'Type=oneshot' cannot have 'Restart=always' or 'Restart=on-success'")
+      ++ (optional hasDeprecated ''Service '${name}.service' uses the attribute
+      StartLimitInterval in the Service section, which is deprecated. See
+      https://github.com/NixOS/nixpkgs/issues/45786.''))
```
I don't know how warnings are rendered, but doesn't using a `''` as done here insert spaces `      ` in the middle of the message ([logic from the manual](https://nixos.org/manual/nix/stable/#idm140737322059824))?

E.g. should this be:

```
... hasDeprecated ''
  Service '${name}.service' uses the attribute
  StartLimitInterval in the Service section, which is deprecated. See
  https://github.com/NixOS/nixpkgs/issues/45786.
''
```

nh2 replied:
I tried it, indeed the formatting is odd this way:

```
trace: warning: Service 'display-manager.service' uses the attribute
      StartLimitInterval in the Service section, which is deprecated. See
      https://github.com/NixOS/nixpkgs/issues/45786.
```

nh2 replied:
Suggestion for the whole `warnings` block:

```nix
    warnings = concatLists (
      mapAttrsToList
        (name: service:
          let
            type = service.serviceConfig.Type or "";
            restart = service.serviceConfig.Restart or "no";
            hasDeprecated = builtins.hasAttr "StartLimitInterval" service.serviceConfig;
          in
            concatLists [
              (optional (type == "oneshot" && (restart == "always" || restart == "on-success"))
                "Service '${name}.service' with 'Type=oneshot' cannot have 'Restart=always' or 'Restart=on-success'"
              )
              (optional hasDeprecated
                "Service '${name}.service' uses the attribute 'StartLimitInterval' in the Service section, which is deprecated. See https://github.com/NixOS/nixpkgs/issues/45786."
              )
            ]
        )
        cfg.services
    );
```

nh2 replied:
I'll force push a refactor that does this in a moment.

nh2 replied:
OK done. Now it renders correctly and the diff looks much nicer.

nh2 committed:
nixos/modules: Reformat `warnings` section

nh2 reviewed (approved):
I've tested your new deprecation warnings on my system, works well. Merging.

nh2 committed:
release notes: Document deprecation warning for StartLimitInterval in [Service]

nh2 commented:
I found some challenges documenting the changes in the release notes, written up in #102246.

roberth commented:
Minimal 20.09 backport for compatibility: https://github.com/NixOS/nixpkgs/pull/112385
